exporter/exporterhelper/batch_sender.go

```diff
@@ -0,0 +1,443 @@
+type BatchMergeFunc func(context.Context, Request, Request) (Request, error)
+
+// BatchSplitFunc is a function that splits a request into multiple requests based on the provided maximum number of
+// items. All the returned requests must have a number of items that does not exceed the maximum number of items.
+// Size of the last returned request must not be more than the size of any other returned request.
+// Do not mutate the request passed to the function if error can be returned after mutation.
+// This API is at the early stage of development and may change without backward compatibility
+// until https://github.com/open-telemetry/opentelemetry-collector/issues/8122 is resolved.
+type BatchSplitFunc func(ctx context.Context, req Request, maxItems int) ([]Request, error)
+
+// IdentifyBatchFunc returns an identifier for a Request batch. This function can be used to separate particular
+// Requests into different batches. Batches with different identifiers will not be merged together.
+// Provided context can be used to extract information from the context and use it as a part of the identifier as well.
+// This function is optional. If not provided, all Requests will be batched together.
+// This API is at the early stage of development and may change without backward compatibility
+// until https://github.com/open-telemetry/opentelemetry-collector/issues/8122 is resolved.
+type IdentifyBatchFunc func(ctx context.Context, r Request) string
+
+type shardGetter interface {
+	shard(*request) (*shard, error)
+}
+
+type Batcher struct {
+	cfg       MergeSplitBatcherConfig
+	mergeFunc BatchMergeFunc
+	splitFunc BatchSplitFunc
+
+	shardGetter
+
+	shutdownCh chan struct{}
+	goroutines sync.WaitGroup
+	logger     *zap.Logger
+
+	export func(*request) error
+}
+
+// shutdown is invoked during service shutdown.
+func (b *Batcher) shutdown() {
+	close(b.shutdownCh)
+
+	// Wait until all goroutines are done.
+	b.goroutines.Wait()
+}
+
+// newShard gets or creates a batcher corresponding with attrs.
+func (b *Batcher) newShard() *shard {
+	s := &shard{
+		batcher:    b,
+		newRequest: make(chan *request, runtime.NumCPU()),
+	}
+	b.goroutines.Add(1)
+	go s.start()
+	return s
+}
+
+func (b *Batcher) mergeRequests(req1 *request, req2 *request) (*request, error) {
+	r, err := b.mergeFunc(req1.Context(), req1.Request, req2.Request)
+	if err != nil {
+		return nil, err
+	}
+	return &request{
+		baseRequest: baseRequest{ctx: req1.Context()},
+		Request:     r,
+	}, nil
+}
+
+func (b *Batcher) splitRequest(req *request, maxItems int) ([]*request, error) {
+	rs, err := b.splitFunc(req.Context(), req.Request, maxItems)
+	if err != nil {
+		return nil, err
+	}
+	reqs := make([]*request, 0, len(rs))
+	for _, r := range rs {
+		reqs = append(reqs, &request{
+			baseRequest: baseRequest{ctx: req.Context()},
+			Request:     r,
+		})
+	}
+	return reqs, nil
+}
+
+func NewMergeBatcher(cfg MergeBatcherConfig, mf BatchMergeFunc, opts ...BatcherOption) *Batcher {
```

Comment:
Instead of passing a function directly I think we should have an interface like it is done for `http.Handler` and `http.HandlerFunc` in the standard library

Reply:
I'm not sure I understand the benefits of using the interfaces here. I don't think we will have a need to add other functions as optional interfaces if that's your concern. Even if we will, it's possible to do with another `BatcherOption`. For me, the function seems to be easier to understand and implement. And we are asking users to provide functions in most of the other exporter helper interfaces.

Reply:
The intent is to be consistent with the standard library, since many users will be familiar with that. I agree it's not particularly likely that we will add other functions, but being similar to the standard library is, I believe, a good goal to have by itself.

Reply:
This is debatable but to me it's also slightly cleaner to write batch merge procedures that keep state if using an interface.

If you have code like this:

```go
type Merger struct {
    state State
}

func (m *Merger) BatchMerge(context.Context, Request, Request) (Request, error) {
     // check or update m.state here
     // ...
}

```

with the current code you have to do:

```go
merger := NewMerger()
mergeBatcher := NewMergeBatcher(cfg, merger.BatchMerge) // merger is passed implicitly
```

while with the interface you directly pass the merger object:

```go
merger := NewMerger()
mergeBatcher := NewMergeBatcher(cfg, merger)
```

Not sure how common this would be, and again, it is debatable, but it seems cleaner with an interface.

Reply:
To not forget @dmitryax please be consistent and apply the same logic https://github.com/open-telemetry/opentelemetry-collector/pull/8764. 

I will think more about this.

Reply:
I personally don't mind any of the solutions, but if we do `Interface` please offer also a `MergerFunc` that helps with implementing the interface simpler. I would also rename `BatchMerge` -> `Merge`.

Reply:
@mx-psi, what kind of `State` do you expect the `Merger` to have? The provided function is expected to be stateless by the API. If we ask users to provide an interface, it makes it not that clear as a stateless function IMO.

Reply:
>  what kind of State do you expect the Merger to have?

A logger or metric instruments could be common examples (potentially also something specific to how you handle requests, but I don't have a concrete example here)

Reply:
I don't think we should encourage users to write logs and emit metrics themselves in the batch function. The function is expected to be slim. All the metrics reporting will be done by the exporter helper. Returned errors should be logged by the exporter helper as well. If there is a real need to do so, it's probably better to keep them in the custom exporter. Then, they will provide `myExporter.mergeBatches` func.

BTW I got feedback from a user of the new interface that having `Export()` function as part of the `Request` interface is pretty inconvenient for the same reason of having to pass a state. Probably, we should consistently move towards functions instead of interfaces so users can provide methods of their custom exporter like `myExporter.convert`, `myExporter.mergeBatches` and `myExporter.export` if needed. The interface would be

```
func NewLogsRequestExporter(
	_ context.Context,
	set exporter.CreateSettings,
	convertFunc ConvertLogsFunc,
	exportFunc ExportFunc,
	options ...Option,
) (exporter.Logs, error)
```

cc @bogdandrutu WDYT?